src/aws_s3_presigned_url.erl

```diff
@@ -0,0 +1,150 @@
+    Client = Client0#{service => <<"s3">>},
+    SecurityToken = maps:get(token, Client, undefined),
+    AccessKeyID = maps:get(access_key_id, Client),
+    Region = maps:get(region, Client),
+    Host = build_host(<<"s3">>, Client, Bucket),
+    {BaseHost, URL0} = build_url(Host, Path, Client, Bucket),
+    Now = calendar:universal_time(),
+    ShortDate = list_to_binary(ec_date:format("Ymd", Now)),
+    Credential = make_credential(AccessKeyID, ShortDate, Region),
+    Headers0 = [{<<"Host">>, Host},
+                {<<"X-Amz-Algorithm">>, <<"AWS4-HMAC-SHA256">>},
+                {<<"X-Amz-Credential">>, Credential},
+                {<<"X-Amz-Date">>, list_to_binary(ec_date:format("YmdTHisZ", Now))},
+                {<<"X-Amz-Expires">>, integer_to_binary(ExpireSeconds)},
+                {<<"X-Amz-SignedHeaders">>, <<"host">>}
+               ],
+    Headers = maybe_add_security_token(Headers0, SecurityToken),
+    Signature = sign_request(Client, MethodBin, BaseHost, aws_request:add_query(URL0, Headers), [{<<"Host">>, BaseHost}], Now, <<"UNSIGNED-PAYLOAD">>),
+    FinalHeaders = lists:keysort(1, [{<<"X-Amz-Signature">>, Signature} | Headers]),
+    {ok, aws_request:add_query(URL0, FinalHeaders)}.
+
+%%====================================================================
+%% Internal functions
+%%====================================================================
+build_host(_EndpointPrefix, #{region := <<"local">>, endpoint := Endpoint}, undefined) ->
+    Endpoint;
+build_host(_EndpointPrefix, #{region := <<"local">>, endpoint := Endpoint}, _Bucket) ->
+    <<Endpoint/binary>>;
+build_host(_EndpointPrefix, #{region := <<"local">>}, undefined) ->
+    "localhost";
+build_host(_EndpointPrefix, #{region := <<"local">>}, _Bucket) ->
+    <<"localhost">>;
+build_host(EndpointPrefix, #{region := Region, endpoint := Endpoint}, undefined) ->
+    aws_util:binary_join([EndpointPrefix, Region, Endpoint], <<".">>);
+build_host(EndpointPrefix, #{region := Region, endpoint := Endpoint}, Bucket) ->
+    aws_util:binary_join([Bucket, EndpointPrefix, Region, Endpoint], <<".">>).
+
+build_url(Host0, Path0, Client, Bucket) ->
+    Proto = maps:get(proto, Client),
+    %% Mocks are notoriously bad with host-style requests, just skip it and use path-style for anything local
+    %% At some points once the mocks catch up, we should remove this ugly hack...
+    Host1 = erlang:iolist_to_binary(Host0),
+    IsLocalHost = maps:get(region, Client) =:= <<"local">>,
+    Path = erlang:iolist_to_binary(Path0),
+    Host = case Bucket of
+             _ when not IsLocalHost andalso Bucket =/= undefined ->
+               erlang:iolist_to_binary(string:replace(Host1, <<Bucket/binary, ".">>, <<"">>, all));
+             _ ->
+              Host1
+           end,
+    Port = maps:get(port, Client),
+    {Host, aws_util:binary_join([Proto, <<"://">>, Host, <<":">>, Port, Path], <<"">>)}.
```

Comment:
Can this be used? https://www.erlang.org/doc/man/uri_string.html#recompose-1

Reply:
Probably, it'd be nice to touch up `aws-codegenerator` at the same time as it does a similar thing. In that case I'd prefer to do it together with the followup of: https://github.com/aws-beam/aws-erlang/pull/125#discussion_r1104155930.